### ISSUE TITLE
`comparison_chain` false positive when constant depends on generic parameter

### ISSUE THREAD

**Issue by clarfonthey** [rust-lang/rust-clippy#13739]
### Description

Constants in patterns cannot depend on generic parameters, so, this should not recommend rewriting into a match when the constants depend on generic parameters.

[Example can be found in my `hetero` crate](https://codeberg.org/clarfonthey/hetero/src/branch/main/src/snoc.rs#L87-L96):

```rust
if idx < I::LEN {
    self.init.get_homo(idx)
} else if idx == I::LEN {
    Some(self.last.into())
} else {
    None
}
```

Lint output:

```
warning: `if` chain can be rewritten with `match`
  --> src/snoc.rs:89:9
   |
89 | /         if idx < I::LEN {
90 | |             self.init.get_homo(idx)
91 | |         } else if idx == I::LEN {
92 | |             Some(self.last.into())
93 | |         } else {
94 | |             None
95 | |         }
   | |_________^
   |
   = help: consider rewriting the `if` chain to use `cmp` and `match`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#comparison_chain
   = note: `#[warn(clippy::comparison_chain)]` on by default
```

If rewritten to a match, like the suggestion implies:

```rust
match idx {
    0..I::LEN => self.init.get_homo(idx),
    I::LEN => Some(self.last.into()),
    _ => None,
}
```

The following error is triggered:

```rust
error[E0158]: constant pattern depends on a generic parameter
  --> src/snoc.rs:90:16
   |
90 |             0..I::LEN => self.init.get_homo(idx),
   |                ^^^^^^

error[E0158]: constant pattern depends on a generic parameter
  --> src/snoc.rs:91:13
   |
91 |             I::LEN => Some(self.last.into()),
   |             ^^^^^^
```

### Version

```text
rustc 1.85.0-nightly (7db7489f9 2024-11-25)
binary: rustc
commit-hash: 7db7489f9bc274cb60c4956bfa56de0185eb1b9b
commit-date: 2024-11-25
host: x86_64-unknown-linux-gnu
release: 1.85.0-nightly
LLVM version: 19.1.4
```

**Comment by y21** [comment]
The lint is suggesting
```rs
match idx.cmp(&I::LEN) {
    Ordering::Less => self.init.get_homo(idx),
    Ordering::Equal => Some(self.last.into()),
    Ordering::Greater => None
}
```
which should work. There's a help message in the diagnostic that mentions to use `cmp`, but I can see how it's easy to miss. Might be a good idea to provide an actual suggestion.
